File: Поиск Фибоначчи.py
def fibonacci_search(arr, target):
    n = len(arr)
    fib2 = 0
    fib1 = 1
    fib = fib2 + fib1
    
    while fib < n:
        fib2 = fib1
        fib1 = fib
        fib = fib2 + fib1
    
    offset = -1
    
    while fib > 1:
        i = min(offset + fib2, n - 1)
        
        if arr[i] < target:
            fib = fib1
            fib1 = fib2
            fib2 = fib - fib1
            offset = i
        elif arr[i] > target:
            fib = fib2
            fib1 = fib1 - fib2
            fib2 = fib - fib1
        else:
            return i
    
    if fib1 == 1 and offset + 1 < n and arr[offset + 1] == target:
        return offset + 1
    
    return -1

arr = [10, 22, 35, 40, 45, 50, 80, 82, 85, 90, 100]
target = 85
result = fibonacci_search(arr, target)
if result != -1:
    print(f"Элемент найден на позиции: {result}")
else:
    print("Элемент не найден")

File: Поиск Фибоначчи.java
public class FibonacciSearch {
    
    public static int findWithFibonacci(int[] numbers, int target) {
        int length = numbers.length;
        
        // Подготовка чисел Фибоначчи
        int fibPrev = 0;      // F(k-2)
        int fibCurrent = 1;   // F(k-1)
        int fibNext = fibPrev + fibCurrent;  // F(k)
        
        // Находим подходящее число Фибоначчи
        while (fibNext < length) {
            fibPrev = fibCurrent;
            fibCurrent = fibNext;
            fibNext = fibPrev + fibCurrent;
        }
        
        int position = -1;
        
        // Поиск в массиве
        while (fibNext > 1) {
            // Вычисляем проверяемый индекс
            int checkIndex = Math.min(position + fibPrev, length - 1);
            
            if (numbers[checkIndex] < target) {
                // Искомый элемент справа
                fibNext = fibCurrent;
                fibCurrent = fibPrev;
                fibPrev = fibNext - fibCurrent;
                position = checkIndex;
            } else if (numbers[checkIndex] > target) {
                // Искомый элемент слева
                fibNext = fibPrev;
                fibCurrent = fibCurrent - fibPrev;
                fibPrev = fibNext - fibCurrent;
            } else {
                // Элемент найден
                return checkIndex;
            }
        }
        
        // Проверка последнего возможного элемента
        if (fibCurrent == 1 && position + 1 < length && 
            numbers[position + 1] == target) {
            return position + 1;
        }
        
        // Элемент не обнаружен
        return -1;
    }
}
